Separate target user

peter1rhodes opened the pull request:
On my work laptop, I have a standard user that I use regularly, but also a local admin account that lets me install stuff. I have therefore modified the install script to allow installation for another user.

I am very open to feedback, as I am a complete Powershell beginner. It is likely that I have not done things in the best way. However, the install script worked well for me

peter1rhodes committed:
First working install.ps1, requires manual username entry in script

peter1rhodes committed:
Untested equivalent mods to uninstall script

peter1rhodes committed:
Default to current user

Default to the current user to preserve standard behaviour

peter1rhodes committed:
Default to current user

Default to the current user to preserve existing behaviour

lextm commented on install.ps1:
```diff
@@ -10,6 +10,19 @@ param(
+
+echo "Changing settings for user $targetUser with SID $targetUserSID"
+
+$Env:LocalAppData = "C:\Users\$targetUser\AppData\Local"
```
Any reason on why `C:` is hard coded here? Not everyone installs Windows on that drive.

peter1rhodes replied:
No that's a great point. I'll have a look for a better way, do you have any suggestions before I look into it?

lextm replied:
You might try to see if `systemroot` environment variable can be used.

lextm commented on uninstall.ps1:
```diff
@@ -9,6 +9,16 @@ param(
+$targetUserObject = New-Object System.Security.Principal.NTAccount($targetUser)
+$targetUserSID = $targetUserObject.Translate([System.Security.Principal.SecurityIdentifier]).value
+
+$Env:LocalAppData = "C:\Users\$targetUser\AppData\Local"
```
Same here with hard coded drive name.

peter1rhodes committed:
Change assumed 'C' drive for :HOMEDRIVE

peter1rhodes committed:
Change hardcoded 'c' for Env:HOMEDRIVE in uninstall.ps1

peter1rhodes committed:
Merge branch 'separate-target-user' of github:peter1rhodes/windowsterminal-shell into separate-target-user
Pulling changes

peter1rhodes commented:
I've opted for $Env:HOMEDRIVE as it seems a clean way to get there

rwasef1830 commented:
@peter1rhodes look up the profile path from registry, some people change the drive that contains the windows profile and home drive may not be the same as the system drive. don't assume all account profiles are in the same drive..

see : HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList

There's probably an easier Win32 call to get this info... or some ready made .NET API.

lextm commented:
`HOMEDRIVE` can be a bad option, as examples like [this](https://itectec.com/superuser/overriding-homedrive-and-homepath-as-a-windows-7-user/) show.

peter1rhodes committed:
Change :HOMEDRIVE for a better way of finding user home directory (Volatile Environment reg keys)

peter1rhodes commented:
Thank you both @rwasef1830  and @lextm. I have found a way to get to the correct user profile directory using HKEY_USERS\[SID]\Voltatile Environment which seems to work well.

peter1rhodes committed:
Rewrite with function to get desired user

peter1rhodes committed:
Change to Get-TargetUserfunction which prompts user

peter1rhodes committed:
Refactor to put Get-TargetUser fn into importable module shared by install and uninstall scripts

peter1rhodes committed:
Merge pull request #1 from peter1rhodes/get-target-user-fn

Get target user fn

peter1rhodes commented:
I've refactored the key parts to use Volatile Environment for both LocalAppData and for the USERPROFILE path. I have also moved all that logic for finding the required paths for the target user to a new function Get-TargetUser, in a new file called TargetUser.psm1, which is then imported by both install.ps1 and uninstall.ps1. The Get-TargetUser function prompts the user to select a user for which to install the context menus, defaulting to the current user if Enter is hit without input. It then returns a custom object containing the required info about the targetUser (their username, SID, LocalAppData path and USERPROFILE path), which can then easily be referenced by the calling scripts.

lextm reviewed (approved):
Looks good, but I don't have a Windows machine at hand to verify. Once I can test them out, I will accept the PR and merge the changes.

AMArostegui commented:
I've tried this PR on my computer and it works

Prodigio commented:
Tried this PR as well, did work for my non-admin account on Windows 10.

EDIT: in order to run this script, one needs to set `Set-ExecutionPolicy -Scope Process -ExecutionPolicy Bypass` in Powershell, otherwise the `TargetUser.psm1` cannot be executed.
